binaries/cuprated/src/rpc/helper.rs

```diff
@@ -0,0 +1,186 @@
+        major_version: header.version.as_u8(),
+        miner_tx_hash: hex::encode(block.miner_transaction.hash()),
+        minor_version: header.vote,
+        nonce: block.header.nonce,
+        num_txes: usize_to_u64(block.transactions.len()),
+        orphan_status,
+        pow_hash,
+        prev_hash: hex::encode(block.header.previous),
+        reward,
+        timestamp: block.header.timestamp,
+        wide_cumulative_difficulty: hex::encode(u128::to_le_bytes(header.cumulative_difficulty)),
+        wide_difficulty,
+    })
+}
+
+/// Same as [`block_header`] but with the block's hash.
+pub(super) async fn block_header_by_hash(
+    state: &mut CupratedRpcHandler,
+    hash: [u8; 32],
+    fill_pow_hash: bool,
+) -> Result<BlockHeader, Error> {
+    let (_, height) = blockchain::find_block(&mut state.blockchain_read, hash)
+        .await?
+        .ok_or_else(|| anyhow!("Block did not exist."))?;
+
+    let block_header = block_header(state, usize_to_u64(height), fill_pow_hash).await?;
+
+    Ok(block_header)
+}
+
+/// Check if `height` is greater than the [`top_height`].
+///
+/// # Errors
+/// This returns the [`top_height`] on [`Ok`] and
+/// returns [`Error`] if `height` is greater than [`top_height`].
+pub(super) async fn check_height(
+    state: &mut CupratedRpcHandler,
+    height: u64,
+) -> Result<u64, Error> {
+    let (top_height, _) = top_height(state).await?;
+
+    if height > top_height {
+        return Err(anyhow!(
+            "Requested block height: {height} greater than top block height: {top_height}",
+        ));
+    }
+
+    Ok(top_height)
+}
+
+/// Parse a hexadecimal [`String`] as a 32-byte hash.
+#[expect(clippy::needless_pass_by_value)]
+pub(super) fn hex_to_hash(hex: String) -> Result<[u8; 32], Error> {
+    let error = || anyhow!("Failed to parse hex representation of hash. Hex = {hex}.");
+
+    let Ok(bytes) = hex::decode(&hex) else {
+        return Err(error());
+    };
+
+    let Ok(hash) = bytes.try_into() else {
+        return Err(error());
+    };
+
+    Ok(hash)
+}
+
+/// [`cuprate_types::blockchain::BlockchainResponse::ChainHeight`] minus 1.
+pub(super) async fn top_height(state: &mut CupratedRpcHandler) -> Result<(u64, [u8; 32]), Error> {
+    let (chain_height, hash) = blockchain::chain_height(&mut state.blockchain_read).await?;
+    let height = chain_height.saturating_sub(1);
+    Ok((height, hash))
+}
+
+/// Check if a key image is spent.
+pub(super) async fn key_image_spent(
+    state: &mut CupratedRpcHandler,
+    key_image: [u8; 32],
+) -> Result<KeyImageSpentStatus, Error> {
+    todo!("impl key image vec check responding KeyImageSpentStatus")
+    // if blockchain::key_image_spent(state, key_image).await? {
+    //     Ok(KeyImageSpentStatus::SpentInBlockchain)
+    // } else if todo!("key image is spent in tx pool") {
+    //     Ok(KeyImageSpentStatus::SpentInPool)
+    // } else {
+    //     Ok(KeyImageSpentStatus::Unspent)
+    // }
+}
```

Comment:
TODO: not JSON-RPC but `/is_key_image_spent` requires something like `BlockchainReadRequest::KeyImagesSpent(Vec<[u8; 32]>) -> BlockchainResponse::KeyImagesSpent(Vec<KeyImageSpentStatus>)`